### DISCUSSION TITLE
Dynamically filter completion

### DISCUSSION THREAD

**Discussion by felipesere** [hrsh7th/nvim-cmp#1851] [Q&A]
Sometimes I get multple hits from my whatever I'm completing, some of the hits a methods, some snippets, some fields and sometimes variables.

I'd love to be able to hit a key stroke and type `Me` and narrow down the currently visible option to just those coming from Methods. The same mechanic could work for variables or fields.

Is this something that exists but I have overlooked?

**Comment by Maykeye** [comment]
It is kinda possible, but the only way I found is a convoluted hack. 
But unless we'll get smart people here, here's a KISS solution.

Plan of attack:

1) Main idea is to use [entry_filter](https://github.com/hrsh7th/nvim-cmp/blob/04e0ca376d6abdbfc8b52180f8ea236cbfddf782/doc/cmp.txt#L646), which AFAIK is the only filter available.
2) Unfortunately it is not dynamic and and is called before list is shown. So... We'll make it behave like if it was 'dynamic'.
3) In config set several keybindings like this

```lua
        mapping = cmp.mapping.preset.insert { ...
          ['<M-q>'] = function()
            cmp.mapping.abort()()
            vim.g.filter_lsp_kind = 'Method'
            cmp.mapping.complete()()
          end,
```

It does three things: Aborts current completion, sets global var to filter only methods(we'll use it), calls completition again(so we can use global var). (I have no idea if there's better way to abort/reopen autocomplete window than to use `mapping.func()()`)

4) Edit source to be like this

```lua
          {
            name = 'nvim_lsp',
            entry_filter = function(entry, ctx)
              if vim.g.filter_lsp_kind ~= nil then
                return require('cmp.types').lsp.CompletionItemKind[entry:get_kind()] == vim.g.filter_lsp_kind
              end
              return true
            end,
          },
```

5) Edit other keys to reset the filter (eg in manual trigger reset filter_lsp_kind to nil, or in confirming it)

Now when you open completition window, press <M-q>. This will reset the filter and reopen completition which will apply the filter and display methods only. 

(Here's example of my [current config](https://gist.github.com/Maykeye/bb6ac04e9a58e30d1f3c1caaa7fd346f) to show how this filterworks )

**Reply by hrsh7th** [reply]
Could you try `cmp.complete` and some temporary settings?

This function is used to "complete once with special settings."

See `:help cmp.complete`

**Reply by Maykeye** [reply]
> Could you try `cmp.complete` and some temporary settings?

Unfortunately I can't find settings on how to filter by kind. I see it's used for [sorting](https://github.com/hrsh7th/nvim-cmp/blob/04e0ca376d6abdbfc8b52180f8ea236cbfddf782/lua/cmp/config/default.lua#L71), but even then can't find how to sort "show Variables above everything else"

Or do you mean store use ConfigSchema instance to store filter_lsp_kind?

**Reply by hrsh7th** [reply]
```
      ['<C-s>'] = cmp.mapping.complete({
        config = {
          sources = {
            {
              name = 'nvim_lsp',
              entry_filter = function(entry)
                return require('cmp.types').lsp.CompletionItemKind[entry:get_kind()] == 'Method'
              end
            }
          }
        }
      }),
```

It seems to be working fine with the above settings, but am I misunderstanding something?

**Reply by Maykeye** [reply]
Oh! Thanks! For some reason I missed that you can do the same filter in sources of config{}! This makes it so much easier
